### PR TITLE
Convert Tick Tank to use DeployToUpgrade + some fixes

### LINKED ISSUE #9242: DeployToUpgrade multiple revocations.
```
Exception of type `System.InvalidOperationException`: Object <OpenRA.Mods.Common.Traits.DeployToUpgrade> revoked more levels of upgrade deployed than it granted for lpst 251.
  at OpenRA.Mods.Common.Traits.UpgradeManager.RevokeUpgrade (OpenRA.Actor self, System.String upgrade, System.Object source) [0x00060] in /Users/thill/projects/openra/OpenRA.Mods.Common//Traits/Upgrades/U
pgradeManager.cs:164
  at OpenRA.Mods.Common.Traits.DeployToUpgrade.RevokeUpgrades () [0x00027] in /Users/thill/projects/openra/OpenRA.Mods.Common//Traits/Upgrades/DeployToUpgrade.cs:168
  at OpenRA.Mods.Common.Traits.WithSpriteBody+<PlayCustomAnimationBackwards>c__AnonStorey3.<>m__0 () [0x00043] in /Users/thill/projects/openra/OpenRA.Mods.Common//Traits/Render/WithSpriteBody.cs:101
  at OpenRA.Graphics.Animation+<PlayThen>c__AnonStorey0.<>m__0 () [0x0008b] in /Users/thill/projects/openra/OpenRA.Game//Graphics/Animation.cs:122
  at OpenRA.Graphics.Animation.Tick (Int32 t) [0x00037] in /Users/thill/projects/openra/OpenRA.Game//Graphics/Animation.cs:182
  at OpenRA.Graphics.Animation.Tick () [0x0001f] in /Users/thill/projects/openra/OpenRA.Game//Graphics/Animation.cs:168
  at OpenRA.Mods.Common.Traits.RenderSprites.Tick (OpenRA.Actor self) [0x00026] in /Users/thill/projects/openra/OpenRA.Mods.Common//Traits/Render/RenderSprites.cs:178
  at OpenRA.World.<Tick>m__3 (TraitPair`1 x) [0x0000e] in /Users/thill/projects/openra/OpenRA.Game//World.cs:286
  at (wrapper delegate-invoke) System.Action`1<OpenRA.TraitPair`1<OpenRA.Traits.ITick>>:invoke_void_T (OpenRA.TraitPair`1<OpenRA.Traits.ITick>)
  at OpenRA.WorldUtils.DoTimed[TraitPair`1] (IEnumerable`1 e, System.Action`1 a, System.String text) [0x00022] in /Users/thill/projects/openra/OpenRA.Game//WorldUtils.cs:53
  at OpenRA.World.Tick () [0x00142] in /Users/thill/projects/openra/OpenRA.Game//World.cs:286
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bf] in /Users/thill/projects/openra/OpenRA.Game//Game.cs:471
  at OpenRA.Game.LogicTick () [0x0004e] in /Users/thill/projects/openra/OpenRA.Game//Game.cs:495
  at OpenRA.Game.Loop () [0x000d4] in /Users/thill/projects/openra/OpenRA.Game//Game.cs:619
  at OpenRA.Game.Run () [0x00042] in /Users/thill/projects/openra/OpenRA.Game//Game.cs:659
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /Users/thill/projects/openra/OpenRA.Game//Support/Program.cs:117
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /Users/thill/projects/openra/OpenRA.Game//Support/Program.cs:39
```

On ff8a578fc9690b5382ba0b5367b07f5b20330a49 of #9241, though there is no reason for that to be relevant.

(cc @penev92)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26419476-deploytoupgrade-multiple-revocations?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #9864: DisableOnUpgrade disables cloak on deployed mobile sensors array (TS)
https://github.com/OpenRA/OpenRA/blob/bleed/mods/ts/rules/shared-vehicles.yaml#L143-L145

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27889039-disableonupgrade-disables-cloak-on-deployed-mobile-sensors-array-ts?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #9996: Crash in the thumper code
```
Dune 2000 Mod at Version playtest-20151114
on map 4da18863d6ae1bec117b6b4163e179a6f73e958a (Tuck's Sietch by Westwood Studios).
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.InvalidOperationException`: Object <OpenRA.Mods.Common.Traits.DeployToUpgrade> revoked more levels of upgrade deployed than it granted for thumper 324.
   at OpenRA.Mods.Common.Traits.UpgradeManager.RevokeUpgrade(Actor self, String upgrade, Object source)
   at OpenRA.Mods.Common.Traits.DeployToUpgrade.RevokeUpgrades()
   at OpenRA.Mods.Common.Traits.DeployToUpgrade.<ResolveOrder>c__AnonStorey1.<>m__0()
   at OpenRA.Activities.CallFunc.Tick(Actor self)
   at OpenRA.Traits.Util.RunActivity(Actor self, Activity act)
   at OpenRA.Actor.Tick()
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Program.Run(String[] args)
   at OpenRA.Program.Main(String[] args)
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28255736-crash-in-the-thumper-code?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by penev92**
Not much to describe here. PR is split into nice self-explanatory commits for your convenience.
~~Fixes an oversight from #9435.~~ (already fixed)
~~Fixes #9242.~~ (already fixed)
~~Fixes #9864.~~ (already fixed)
~~Fixes #9996.~~ (already fixed)

### CODE REVIEW

**Comment by abcdefg30** [comment]
> OpenRA.Utility(1,1): Error: Actor ttnk using voice set Vehicle does not define Action voice required by OpenRA.Mods.Common.Traits.AttackTurretedInfo

**Comment by reaperrr** [comment]
You probably need to set `Voice: Attack` on `AttackTurreted`.

**Comment by penev92** [comment]
Fixed the voice.

**Comment by atlimit8** [comment]
`DeployingUpgrade` should not be active while deployed since deploying means getting deployed - not deployed.
You probably should add an `UndeployedUpgrade` to enable traits while fully undeployed. 

In fact, since `DeployedUpgrades` and `UndeployedUpgrade` could be used together to enable/disable traits while in transition as well, `DeployingUpgrade` could be entirely replaced by `UndeployedUpgrade`.

**Comment by penev92** [comment]
>DeployingUpgrade should not be active while deployed since deploying means getting deployed - not deployed.

True, but that makes for some ugly and redundant YAML, so I decided there wasn't any harm in keeping it.

>You probably should add an UndeployedUpgrade to enable traits while fully undeployed. 

That sounds like taking the whole "everything is an upgrade" idea to a whole new level, which doesn't sound like a something we'd want to start doing on a large scale.

**Comment by atlimit8** [comment]
If you are against adding `UndeployedUpgrade` then `DeployingUpgrade` should be renamed `NotUndeployedUpgrade` since "deploying" is incorrect.

On another note, "upgrades" should be renamed "conditions" this weekend, which should help with confusion on the matter especially when condition expression traits such as `ConditionCombiner` are added.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by penev92** [comment]
Rebased.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ public Attack(Actor self, Target target, bool allowMovement, bool forceAttack)
 
 			this.forceAttack = forceAttack;
 
-			attack = self.Trait<AttackBase>();
+			attack = self.TraitsImplementing<AttackBase>().First(x => !x.IsTraitDisabled);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: What if that specific AttackBase becomes disabled between the constructor and tick (which may be quite far apart if the activity is queued)?

Comment (atlimit8) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Good point.
Perhaps `attack` should be initialized (`attack = self.TraitsImplementing<AttackBase>().FirstOrDefault(x => !x.IsTraitDisabled)`) on first `Tick` and the activity canceled if `attack == null`.

Comment (penev92) [inline OpenRA.Mods.Common/Activities/Attack.cs]: More like that should be fetched on every tick, which is a bit ridiculous.

Comment (atlimit8) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Actually, all that needs to be done in other ticks is check if cached trait is disabled and cancel if it is.
```c#
 		public override Activity Tick(Actor self)
 		{
			if (attack == null)
			{
				attack = self.TraitsImplementing<AttackBase>().FirstOrDefault(x => !x.IsTraitDisabled)
				if (attack == null)
					return NextActivity;
			}
			else if (attack.IsTraitDisabled)
				return NextActivity;

 			var ret = InnerTick(self, attack);
 			attack.IsAttacking = ret == this;
 			return ret;
 		}
```

Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: If you store the IEnumerable<AttackBase> then the filtering each tick will be relatively cheap.

**Comment by penev92** [comment]
This now depends on #10014, so I'll close it for the moment.

**Comment by penev92** [comment]
Revived.

**Comment by abcdefg30** [comment]
Looks good to me. :+1: 

But this needs a) a solution for the importer and b) a rebase.

**Comment by penev92** [comment]
Done.

**Comment by Mailaender** [comment]
The tick tank will stop firing when deploying/undeploying. This isn't a regression though.

File: mods/ts/rules/nod-vehicles.yaml [hunk 2]
```diff
@@ -100,12 +100,67 @@ TTNK:
...
 	RevealsShroud:
 		Range: 5c0
-	Transforms:
-		IntoActor: gatick
-		Facing: 159
-		TransformSounds: place2.aud
-		NoTransformSounds:
-		Voice: Move
+	RenderSprites:
+		Image: ttnk
```
Comment (Mailaender) [inline mods/ts/rules/nod-vehicles.yaml]: Isn't this redundant?

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c72403dfcee9a3e220625e1091c6f4686f35caa1...bfd4dc37658fb384985f4805a5042ac95713b342)

**Pull request merged**